### ISSUE TITLE
Regression related to forward reference and nested classes

### ISSUE THREAD

**Issue by JukkaL** [python/mypy#18988]
This example started generating a false positive about an undefined name (simplified from a real-world generated protobuf stubs):
```py
from typing import NewType

class W[T]: pass

class R:
    class M(W[Action.V], type):  # Name "Action.V" is not defined
        FOO = R.Action.V(0)
    class Action(metaclass=M):
        V = NewType('V', int)

class Action:
    pass
```

This was a side effect of #18625.

There may be some confusion between `R.Action` and `Action` classes. 

This looks like a release blocker. This is happening in generated protobuf stubs code which can't be edited easily. 

If this is non-trivial to fix, we could temporarily revert the original PR until a fix is available.

cc @ilevkivskyi as the author of the above PR

**Comment by ilevkivskyi** [comment]
A possible quick fix may be to reduce precedence of global forward reference (although this may further complicate precedence rules for forward references). I can make a PR later today and check what is the fallout (likely there will be none, so it is more a question of documenting this clearly at some point).

**Comment by sterliakov** [comment]
Line 6 in your example produces a runtime `NameError` (verified on 3.12, 3.13, and my common sense), so if anything is to be changed here, that should only apply to stubs?..

**Comment by JukkaL** [comment]
There could also be a string literal forward reference in a non-stub file. The main thing is maintaining backward compatibility.

**Comment by ilevkivskyi** [comment]
Here is a possible fix https://github.com/python/mypy/pull/19000
